Using KaTeX in docs

TianyiShi2001 opened the pull request:
Using KaTeX in docs, as suggested in isuue #349 

For local preview, use `RUSTDOCFLAGS="--html-in-header katex-header.html" cargo doc`

<img width="1164" alt="スクリーンショット 2020-09-07 9 07 04" src="https://user-images.githubusercontent.com/38644266/92339614-a4229f00-f0e9-11ea-83c2-947c05fb6032.png">

dlaehnemann commented:
I guess it's just a demo for now. I like it. Eventually, this PR should then probably adjust this across the whole repo, to keep the docs  uniform.

TianyiShi2001 commented:
I've updated all math expressions where using LaTeX is appropriate.

In addition, I updated the documentation guideline regarding the format of variable names and math expressions (when to use backticks and when to use LaTeX).

dlaehnemann commented on README.md:
```diff
@@ -50,6 +50,46 @@ For extra credit, feel free to familiarize yourself with:
 * the Rust [documentation conventions](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text)
 * the Rust [API documentation guidelines](https://rust-lang.github.io/api-guidelines/documentation.html)
 
+#### About variable names and mathematical expressions
```
I would skip the `About` here, to shorten a bit:
```suggestion
#### variable names and mathematical expressions
```

dlaehnemann commented on README.md:
```diff
@@ -50,6 +50,46 @@ For extra credit, feel free to familiarize yourself with:
 
+#### About variable names and mathematical expressions
+
+* When referring to a variable as an arguments of the function or a field of the struct, put backticks (`` ` ``) around it.
+* Otherwise, put dollar signs (`$`) around it (i.e. making it a LaTeX expression).
+* In general, use LaTeX for math expressions (especially when writing "big-O" complexities)
+* For simple expressions, especially those containing function argument/struct field names, using backticks is recommended.
```
Merging the first and last row, and the second and third row for brevity:
```suggestion
* For mathematical expressions in doc strings (e.g. for "big-O" notation), you can use LaTeX syntax in between dollar signs (e.g. `$O(n)$`) in-line.
* For simple expressions that refer to a variable as an argument of a function or to a field of a struct, using backticks (`\`field_1 and var_x\``) is recommended.
```

dlaehnemann commented on README.md:
````diff
@@ -50,6 +50,46 @@ For extra credit, feel free to familiarize yourself with:
+* In general, use LaTeX for math expressions (especially when writing "big-O" complexities)
+* For simple expressions, especially those containing function argument/struct field names, using backticks is recommended.
+
+In the following example, the public function `combinations` has two arguments `n` and `k`, which are referred to in the first line with backticks. Then, a LaTeX math expressions is used to show the complexity. The variable names *k* and *n* are not ambiguous (obviously referring to the two arguments), so we don't need to define again.
+
+```rust
+/// Calculate the number of combinations when choosing
+/// `k` elements from `n` elements without replacement.
+///
+/// Time complexity: $O(\min(k, n - k))$
+pub fn combinations(n: u64, k: u64) -> f64 {
+    scaled_combinations(n, k, 1.0)
+}
+```
+
+In the next example, none of the variables *n*, *k* and *w* used in the space complexity expression matches any field names of the struct `BitEnc`, so we need to define explicitly. Simple expressions like `width = 7` and `4 * 7 = 28` are enclosed in backticks.
+
+```rust
+/// A sequence of bitencoded values.
+///
+/// Space complexity: $O(\dfrac{nw}{k})$; specifically $\Big\lceil\dfrac{nw}{k}\Big\rceil \times 32$
+/// bits, where $w$ is `width`, $n$ is the length of the input
+/// sequence, and $k = 32 - (32 \mod w)$ is the number of bits in each
+/// 32-bit block that can be used to store values.
+/// For values that are not a divider of 32, some bits will remain unused.
+/// For example for `width = 7` only `4 * 7 = 28` bits are used.
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct BitEnc {
+    storage: Vec<u32>,
+    width: usize,
+    mask: u32,
+    len: usize,
+    usable_bits_per_block: usize,
+}
````
These examples are great, but a bit lengthy. Maybe we can either shorten them or replace them with links to the respective code, so that users can have a look at them for good templates?

TianyiShi2001 replied:
How about making a separate `CONTRIBUTING.md` to avoid cluttering `README.md`, while not losing information? Variable formatting conventions aren't something I would expect in  `README.md` in the first place. In the future there may be more guidelines, such as "where to put references" and "how to format references"

dlaehnemann commented on src/stats/pairhmm/mod.rs:
```diff
@@ -102,8 +102,8 @@ pub use pairhmm::PairHMM;
 
-mod homopolypairhmm;
-mod pairhmm;
+pub mod homopolypairhmm;
+pub mod pairhmm;
```
Is this change intended for this pull request? Seems off-topic and if this was not explicitly `pub` before, it should probably stay that way?

TianyiShi2001 replied:
I was checking the correctness of the output but private modules were not included by default when building the docs, so I made them temporarily public but forgot to restore them. Actually I thought it would be helpful to document items that are currently private. But this should be a separate issue. I've restored them as private modules.

dlaehnemann reviewed (commented):
Very cool effort, thanks!

I have some suggestions on the `README.md` changes and a question about making two submodules `pub`.

pmarks commented:
The Katex math is really nice! Unfortunately, my understanding is that this will not be supported on [docs.rs](docs.rs) anytime soon ([see disucssion](https://github.com/rust-num/num/pull/226)), so people accessing the docs that way (or just running `cargo doc` without being aware of this will see the raw Katex strings.  So it's a bit of a tradeoff until Katex is more fully supported by rustdoc.  Just wanted to flag that point for discussion.

dlaehnemann commented:
Hmm, I was under the impression that this is merely using the existing `rustdoc` command-line arg `--html-in-header` [in a way that doc.rs allows](https://docs.rs/about/metadata):
https://github.com/rust-bio/rust-bio/pull/352/files#diff-80398c5faae3c069e4e6aa2ed11b28c0R68

And then it simply injects the java script needed for the rendering from externally:
https://github.com/rust-bio/rust-bio/pull/352/files#diff-d648f2b0571abb5774532138b8b4d62fR1-R15

But maybe I'm missing something here, and rustdoc will somehow catch this injection and prevent it?

pmarks commented:
@dlaehnemann  ah, interesting, maybe it will work! it's too bad that the extra rustdoc args don't seem to be injected into a local run of `cargo doc`

dlaehnemann commented:
From what I understand, docs.rs runs `cargo run -- build` to compile the docs and supposedly you can do this pointing to some crate in your local file system:
https://github.com/rust-lang/docs.rs/blob/master/README.md#build-subcommand

But the docs.rs documentation is surprisingly sparse. But it links to the above (hyperlinks are not highlighted in any way, but are the default text color...), here ("Testing documentation builds locally"):
https://docs.rs/about/builds

dlaehnemann commented:
@TianyiShi2001 Do you happen to know an example crate where this use of KaTeX works? Would be nice to have as proof that it actually works on docs.rs before we merge this.... :D

TianyiShi2001 commented:
> @TianyiShi2001 Do you happen to know an example crate where this use of KaTeX works? Would be nice to have as proof that it actually works on docs.rs before we merge this.... :D

@dlaehnemann yes, here: https://docs.rs/rustdoc-katex-demo/0.1.5/rustdoc_katex_demo/

dlaehnemann commented:
Thanks @TianyiShi2001 , I had seen this earlier but forgotten where...

@pmarks I guess this proves that this works right now. So unless we don't want to call external JS stuff, this is good to go?

For some reason, the CI coverage calculations run forever and then simply get cancelled, same over at PR #354. I'm not sure why. Anybody any ideas?

pmarks commented:
@TianyiShi2001 @dlaehnemann very cool that this works on docs.rs!! Thanks!

alxs commented:
I just randomly stumbled upon this and I think it deserves a ping :D

dlaehnemann commented:
Absolutely. I'll try and see this PR through, today. I'll resolve the one point that is outstanding, to create a separate `Contribute.md` where we can move detailed example HowTos.

dlaehnemann commented:
@TianyiShi2001 Do these changes work for you? I'll merge the PR once you agree.
